keps/sig-apps/2307-job-tracking-wihout-lingering-pods/README.md

```diff
@@ -0,0 +1,601 @@
+    // been accounted in the counters.
+    UncountedPodUIDs UncountedPodUIDs
+}
+
+// UncountedPodUIDs holds UIDs of Pods that have finished but haven't been
+// accounted in Job status counters.
+type UncountedPodUIDs struct {
+    // Enabled indicates whether Job tracking uses this struct.
+    Enabled   bool
+    // Succeeded holds UIDs of succeeded Pods.
+    Succeeded []string
+    // Succeeded holds UIDs of failed Pods.
+    Failed    []string
+}
+```
+
+### Algorithm
+
+The following algorithm updates the status counters without relying on finished
+Pods to be present indefinitely. The algorithm assumes that the Job controller
+could be stopped at any point and executed again from the first step without
+losing information. Generally, all the steps happen in a single Job sync
+cycle.
+
+1. The Job controller creates missing Pods with the finalizer
+   `batch.kubernetes.io/job-completion`
+2. The Job controller adds Pod UIDs to the `.status.uncountedPodUIDs.succeeded`
+   and `.status.uncountedPodsUIDs.failed` lists if the Pod:
+    - has the `batch.kubernetes.io/job-completion` finalizer, and
+    - the Pod is on Succeeded or Failed phase, respectively.
+    The controller sends a status update.
+3. The Job controller removes the `batch.kubernetes.io/job-completion` finalizer
+   from all Pods on Succeeded or Failed phase. We can assume that all of them
+   have been added to the lists if the previous step succeeded. The controller
+   keeps note of the Pod updates that succedeed.
+4. The Job controller counts the Pods in the `.status.uncountedPodUIDs` lists
+   that:
+   - had no finalizer at the beginning of the sync, or
+   - had the patch in step 3 succeed, or
+   - were removed from the system.
+   The counts increment the `.status.failed` and `.status.succeeded` and clears
+    counted Pods from `.status.uncountedPodsUIDs` lists. The controller sends a
+    status update.
+
+Steps 2 to 4 might deal with a potentially big number of Pods. Thus, status
+updates can potentially stress the kube-apiserver. For this reason, the Job
+controller repeats steps 2 to 4, dealing with up to 100 Pods each time, until
+the controller processes all the Pods.
+
+### User deleted Pods
+   
+In the case where a user or another controller removes a Pod, which sets a
+deletion timestamp, the Job controller removes the finalizer from the Pod,
+<<[UNRESOLVED]>>
+independently of it having finished or not.
+This allows for deleted Pods to not count towards the failed counter, like it
+is today.
+Note that step 2 doesn't pay attention to deletion timestamps. So the Job
+controller might account the Pod in the lists and counters if the Pods
+enter Failed or Succeeded phase before or while being deleted.
+<<[/UNRESOLVED]>>
+   
+### Pod adoption
+
+<<[UNRESOLVED Pod adoption ]>>
+If a Job with `.status.uncountedPodUIDs.enabled = true` can adopt a Pod, this
+might not have a finalizer. We can:
```

Comment:
The qualifications for Pod adoption don't change from current behavior. Clarified.

Reply:
Does the current qualification criteria include pods that failed/succeeded? If the pod was in finished/succeeded status, the absence of the finalizer could be because the pod was already processed by the controller in a previous cycle.

Reply:
In that scenario, the Pod would already have an updated owner reference.

Reply:
what guarantees that? is it possible that an owner reference be cleared after it was counted for?

Reply:
Actually there is currently no criteria other than the Job selector, which is mutable.

In theory, a Pod could finish, be accounted for Job A, loose its finalizer and then be claimed by another Job B. At this point, I think it's fair to adopt this Pod and account for it in Job B.
But what if Job A claims the Pod again? Then we have no way of adopting it without counting it again.

@wojtek-t @lavalamp thoughts?

Perhaps we should actually not adopt a Pod if it doesn't have the finalizer.

Reply:
You should really not have this problem, because only one job can have the "controller reference"; you should not clear that, and no one else should either.

Reply:
right, users should not, but they could. While unlikely, I think this is worth a note because in practice this is a regression compared to current situation.

Reply:
Are you referring to this being documented as unexpected behavior? https://kubernetes.io/docs/concepts/workloads/controllers/job/#specifying-your-own-pod-selector

Although I'm proposing an scenario where ownership is transferred from Job A to Job B and then back to Job A by carefully updating the selectors so that there are no conflicts at a given time. But maybe we can document this as unexpected behavior too.